src/rust/session/src/buffer.rs

```diff
@@ -75,35 +79,47 @@ impl Buffer {
     /// Mark that `amt` bytes have been consumed and should not be returned in
     /// future reads from the buffer.
     pub fn consume(&mut self, amt: usize) {
+        let old_capacity = self.buffer.capacity();
         self.read_offset = std::cmp::min(self.read_offset + amt, self.write_offset);
         if self.is_empty() {
             // if the buffer is empty, we can simply shrink it down and move the
             // offsets to the start of the buffer storage
             self.write_offset = 0;
             self.read_offset = 0;
-            if self.buffer.len() > self.target_capacity {
+            if self.buffer.len() - self.target_capacity > 0 {
+                // buffer can be reduced back down to the target
                 self.buffer.truncate(self.target_capacity);
                 self.buffer.shrink_to_fit();
             }
-        } else if self.read_offset > self.target_capacity {
-            // this case results in a memmove of the buffer contents to the
-            // beginning of the buffer storage and tries to free additional
-            // space
+        } else if self.buffer.len() > self.target_capacity && self.len() * 2 < self.buffer.len() {
+            // the buffer is both oversized and less than half full, we can
+            // shrink the buffer size after memmove-ing the contents to the
+            // beginning of the buffer
             self.buffer
                 .copy_within(self.read_offset..self.write_offset, 0);
             self.write_offset -= self.read_offset;
             self.read_offset = 0;
-            // if the buffer is occupying less than 1/2 of the storage capacity
-            // we can resize it to free up the unused space at the end of the
-            // buffer storage.
-            if self.len() < self.buffer.capacity() / 2 {
-                self.buffer
-                    .truncate(std::cmp::max(self.len(), self.target_capacity));
-                // SAFETY: we maintain our own write offset which ensures the
-                // uninitialized bytes are not read until they have been written.
-                unsafe {
-                    self.buffer.set_len(self.buffer.capacity());
-                }
+
+            let target = self.buffer.len() / 2;
+
+            self.buffer.truncate(target);
+            self.buffer.shrink_to_fit();
+        }
+
+        // update stats if the buffer has resized
```

Comment:
This seems more verbose than duplicating the `Less` condition in two places above.

Reply:
Consolidated some of the logic above, see if this looks better now.